libgalois/include/katana/PropertyGraph.h

```diff
@@ -170,7 +170,7 @@ class KATANA_EXPORT PropertyGraph {
     Result<void> (PropertyGraph::*add_properties_fn)(
         const std::shared_ptr<arrow::Table>& props);
     Result<void> (PropertyGraph::*upsert_properties_fn)(
-        const std::shared_ptr<arrow::Table>& props);
+        const std::shared_ptr<arrow::Table>& props, tsuba::TxnContext& ctx);
```

Comment:
We prefer C pointers to mutable references.  Is there a compelling case for a mutable reference here?

Reply:
Hmm... I saw other contexts mostly all uses references. e.g. `ThreadContext` is also mutable and uses a mutable reference. So I thought it's a convention.

After a second thought, I think pointer may make more sense. So I'm going to change all of them into pointers if you agree on this?

Reply:
I read that some people tend to use references unless NULL value of the pointer is allowed. "Since references can’t be NULL, they are safer to use." I think we don't allow a NULL pointer to `TxnContext` in our case, and operations outside of transactions are also treated as single operation transactions, hence it's not necessary to do NULL checking on the potential `TxnContext` pointer. Maybe pass by reference is better?

Reply:
@ddn0, language feature consult please.

Reply:
@ddn0 any opinion on this? Thanks!

Reply:
The traditional arguments are:
1. As Yige brings up, references are better than pointers because they express non-nullablity
2. Pointers are better than references because callers explicitly opt-in to mutation by using `&`

I'm pretty sure our codebase is not consistent about this either, but our style guide does prefer (2) over (1): https://docs.k9h.dev/0.1.101.dev+51.303.ac3e9a.4018d6/contributing/code-style.html#code-style

And we do have clang-tidy rules that warn on mutable references.

Since one cannot do both (1) and (2), conventions that make up the difference.

If you are in the world of (1), you have the expectation that callers will pay attention to the signature of functions if they care about mutations of parameters ("people should read the docs!").

If you are in the world of (2), you have the expectation that either (a) a function accepting a null pointer is documented ("people should read the docs!") or objects of a certain type always exist ("don't be a monster").

I'd suggest the latter for us. That is, even if there is no transaction, we should have a dummy transaction context and all callees can assume that it is safe to dereference the transaction context.

Reply:
Thanks very much.
When the rules and logic conflict, it is tough to know what to do.
I am a fan of C-style pointers for mutable arguments.
And `KATANA_LOG_DEBUG_ASSERT(txn != nullptr);` provides a zero-cost documentation to programmers that you expect a valid transaction context.